src/jllvm/materialization/CodeGeneratorUtils.cpp

```diff
@@ -77,8 +77,394 @@ class TrivialDebugInfoBuilder
+            { typeStack.erase(typeStack.end() - 3, typeStack.end()); },
+            [&](OneOfBase<AConstNull, ALoad, ALoad0, ALoad1, ALoad2, ALoad3, New>)
+            { typeStack.push_back(m_addressType); },
+            [&](OneOfBase<AReturn, AThrow, DReturn, FReturn, IReturn, LReturn, Return>) { done = true; },
+            [&](OneOfBase<AStore, AStore0, AStore1, AStore2, AStore3, DAdd, DDiv, DMul, DRem, DStore, DStore0, DStore1,
+                          DStore2, DStore3, DSub, FAdd, FDiv, FMul, FRem, FStore, FStore0, FStore1, FStore2, FStore3,
+                          FSub, IAdd, IAnd, IDiv, IMul, IOr, IRem, IShl, IShr, IStore, IStore0, IStore1, IStore2,
+                          IStore3, ISub, IUShr, IXor, LAdd, LAnd, LDiv, LMul, LOr, LRem, LShl, LShr, LStore, LStore0,
+                          LStore1, LStore2, LStore3, LSub, LUShr, LXor, MonitorEnter, MonitorExit, Pop, PutStatic>)
+            { typeStack.pop_back(); },
+            [&](OneOfBase<ArrayLength, D2I, F2I, InstanceOf, L2I>) { typeStack.back() = m_intType; },
+            [&](OneOfBase<CheckCast, DNeg, FNeg, I2B, I2C, I2S, IInc, INeg, LNeg, Nop>) { /* Types do not change */ },
+            [&](OneOfBase<BALoad, CALoad, DCmpG, DCmpL, FCmpG, FCmpL, IALoad, LCmp, SALoad>)
+            {
+                typeStack.pop_back();
+                typeStack.back() = m_intType;
+            },
+            [&](OneOfBase<BIPush, IConstM1, IConst0, IConst1, IConst2, IConst3, IConst4, IConst5, ILoad, ILoad0, ILoad1,
+                          ILoad2, ILoad3, SIPush>) { typeStack.push_back(m_intType); },
+            [&](OneOfBase<D2F, I2F, L2F, FALoad>)
+            {
+                if (holds_alternative<FALoad>(operation))
+                {
+                    typeStack.pop_back();
+                }
+                typeStack.back() = m_floatType;
+            },
+            [&](OneOfBase<D2L, F2L, I2L, LALoad>)
+            {
+                if (holds_alternative<LALoad>(operation))
+                {
+                    typeStack.pop_back();
+                }
+                typeStack.back() = m_longType;
+            },
+            [&](OneOfBase<DALoad, F2D, I2D, L2D>)
+            {
+                if (holds_alternative<DALoad>(operation))
+                {
+                    typeStack.pop_back();
+                }
+                typeStack.back() = m_doubleType;
+            },
+            [&](OneOfBase<DConst0, DConst1, DLoad, DLoad0, DLoad1, DLoad2, DLoad3>)
+            { typeStack.push_back(m_doubleType); },
+            [&](Dup) { typeStack.push_back(typeStack.back()); },
+            [&](DupX1)
+            {
+                auto iter = typeStack.end();
+                llvm::Type* type1 = *--iter;
+                llvm::Type* type2 = *--iter;
+
+                assert(!isCategoryTwo(type1) && !isCategoryTwo(type2));
+
+                typeStack.insert(iter, type1);
```

Comment:
I think these might read nicer by using the reverse iterators.

Here you could e.g. write 
```cpp
auto iter = typeStack.rbegin();
llvm::Type* type1 = *iter++;
llvm::Type* type2 = *iter++;
```
which is a more familiar pattern.

Ditto below

Reply:
`insert` does not work with reverse iterators

Reply:
You can use `.base()` on the revese iterator to get the underlying iterator that can be used in `insert`